atomicapp/providers/kubernetes.py

```diff
@@ -38,227 +44,243 @@ class KubernetesProvider(Provider):
-                else:
-                    os.symlink(host_kube_conf_path, kube_conf_path)
-        else:
-            self.kubectl = self._find_kubectl()
+
+        self._process_artifacts()
 
         if not self.dryrun:
-            if not os.access(self.kubectl, os.X_OK):
-                raise ProviderFailedException("Command: " + self.kubectl + " not found")
 
-            # Check if Kubernetes config file is accessible, but only
-            # if one was provided by the user; config file is optional.
-            if self.config_file:
-                self.checkConfigFile()
+            '''
+            Config_file:
+                If a config_file has been provided, use the configuration
+                from the file and load the associated generated file.
+                If a config_file exists (--provider-config) use that.
 
-    def _find_kubectl(self, prefix=""):
-        """Determine the path to the kubectl program on the host.
-        1) Check the config for a provider_cli in the general section
-           remember to add /host prefix
-        2) Search /usr/bin:/usr/local/bin
+            Params:
+                If any provider specific parameters have been provided,
+                load the configuration through the answers.conf file
 
-        Use the first valid value found
-        """
+            .kube/config doesn't exist:
+                If the configuration file does not exist, generate one using
+                the default localhost:8080/api end-point.
+
+            Default configuration:
+                If no paramaters have been provided, as well as no --provider-config,
+                use the default .kube/config location and load the associated configuration.
+            '''
 
-        if self.dryrun:
-            # Testing env does not have kubectl in it
-            return "/usr/bin/kubectl"
+            config_location = os.path.join(Utils.getRoot(), Utils.getUserHome().strip('/'), '.kube/config')
 
-        test_paths = ['/usr/bin/kubectl', '/usr/local/bin/kubectl']
-        if self.config.get("provider_cli"):
-            logger.info("caller gave provider_cli: " + self.config.get("provider_cli"))
-            test_paths.insert(0, self.config.get("provider_cli"))
+            if self.config_file:
+                self.api = Client(KubeConfig.from_file(self.config_file), "kubernetes")
+            elif self._check_required_params():
+                self.api = Client(self._from_required_params(), "kubernetes")
+            elif not os.path.isfile(config_location):
+                self.config["provider-api"] = self.provider_api
+                self.api = Client(self._from_required_params(), "kubernetes")
+            else:
+                self.api = Client(KubeConfig.from_file(config_location), "kubernetes")
+
+            # Check if the namespace that the app is being deployed to is available
+            self._check_namespaces()
+
+    def _check_required_params(self):
```

Comment:
btw this is a great example of a function that can be implemented for kubernetes and openshift - since they will probably have different requirements of what is required. openshift could inherit from a library and then "override" this function with one of its own.